json-everything.net/wwwroot/md/vocabs-data-2022.md

```diff
@@ -0,0 +1,123 @@
+
+## 3. The `data` Keyword
+
+### 3.1 Syntax and Semantics
+
+The value of `data` must be an object.  The keys of the object correspond to valid JSON Schema keywords, and the values MUST be one of
+
+- JSON Pointers per [RFC 6901](https://www.rfc-editor.org/rfc/rfc6901)
+- [Relative JSON Pointers](https://json-schema.org/draft/2019-09/relative-json-pointer.html)
+- URI references per [RFC 3986](https://datatracker.ietf.org/doc/html/rfc3986), potentially with JSON Pointer fragments
+
+The set of keywords in the schema or subschema containing `data` SHOULD be distinct from the set of keys defined within `data`.  Behavior in the event of an overlap is not defined.
+
+`data` operates in two phases:
+
+1. All of the values are dereferenced per sections 3.2 and 3.3.
+2. The keywords and values then form a schema which is applied to the instance at the current location.
+
+The validation and annotation results of `data` are those of the formed schema.  More detail regarding output can be found in section 3.4.
+
+### 3.2 Contextual Behavior
+
+`data` MUST be processed contextually in accordance with the draft of the schema in which it is used.  For example, if `data` is used in a schema that declares draft 2019-09, then
+
+- the keys SHOULD be recognized by draft 2019-09 or defined in other referenced vocabularies, and
+- the URI references MUST be resolved following the same rules as `$ref` as defined by draft 2019-09.
+
+If a key is not a recognized keyword in the schema's draft (e.g. `$dynamicAnchor` in a draft 2019-09 schema), then this key is ignored just as it would be ignored if it were in the schema or subschema containing `data`.
+
+### 3.3 URI Resolution
+
+The values of `data` are dereferenced in different ways depending on the format of the value.
+
+If the value is a JSON Pointer, it is resolved against the instance root.
+
+If the value is a Relative JSON Pointer, it is resolved against the instance at the location currently being evaluated.
+
+If the value is a URI reference, it must be resolved in accordance with the rules of `$ref` resolution for the containing spec, as per section 3.2.  The primary different between resolving URIs for `data` and resolving URIs for `$ref` is that `$ref` requires that the indicated data must represent a valid schema object whereas `data` can reference any valid JSON value.
+
+External document data can be referenced using an absolute URI base identifier along with a JSON Pointer fragment.
+
+Implementations SHOULD be able to download external JSON documents, but MAY provide means to pre-load and cache such documents against the identifiers by which they will be referenced.  Documents downloaded from URIs which contain a JSON Pointer fragment MUST be interpreted using a media type that allows resolution of such fragments.
```

Comment:
I forgot something:  It would better align with JSON Schema's policies on remote resources in general, and `$ref` resolution in particular, if you switched this so that the pre-load cache is a SHOULD while downloading external documents is a MAY.

It's also not realistic to expect an implementation to support all possible URL schemes for downloads, plus URN-like URI schemes can't be downloaded anyway.

Reply:
I went with this primarily because a lot of the original discussion was around fetching data from external sources, possibly unknown until evaluation time.  Aligning with the spec is a good argument, but I do want to leave room for fetching data.

Reply:
I think the only distinction here is whether the application fetches the data immediately beforehand and passes it to the implementation (somehow... we don't have a notion of populating a non-schema data store as part of the implementation interface), vs actually re-doing the fetch every time the `data` keyword is hit.

How important is it to re-do the fetch every time?

Anyway, I'm not dead-set against fetching, it just feels a little outside of JSON Schema norms.  Then again, I'm not sure how a vocabulary can require a pre-populatable data cache, so... 🤔